README.md
        + ======================================================================================== +
        |                              [  Snake Class Methods  ]                                   |
        + ======================================================================================== +

    1)   snake_body()
            It Creates Initial snake body with Default position
            with Head and 3 Tail Segment

    2)   add_segment()
           +----------------------+---------------+
           | Parameter of Method  | Default Value |
           +----------------------+---------------+
           |     position         |  user_define  |
           |     shape            |  Square       |
           |     color            |  Black        |
           |     length           |  0.5          |
           |     width            |  0.5          |
           +----------------------+---------------+
           It Creates New Segment and Make Object of the Segment and,
           It Stores in snake_segment List

    3)  extend()
           It Creates New Segment at last and, at the position last segment
                (It helps to extend snake when snake eat food)

    4)  move()
           It Moves whole Snake Forward and, [ It Returns Position of Snake head ]

    5)  up()
           It Turns Head Up[ 90 ] degree

    6)  down()
           It Turns Head Down[ 270 ] degree

    7)  right()
           It Turns Head Right[ 0 ] degree

    8)  left()
           It Turns Head left[ 180 ] degree

    9)  reset()
           It Reset all values to default or Set all at stating point

    10) snake_boundary()
           +----------------------+---------------+
           | Parameter of Method  | Default Value |
           +----------------------+---------------+
           |     head_position    |  user_define  |
           |     x_positive       |  300          |
           |     x_negative       |  -300         |
           |     y_positive       |  300          |
           |     y_negative       |  -300         |
           +----------------------+---------------+
           It Set End Boundary of Snake at Position Snake can Move
           and , Return Boolean Value


  ========================================================================================================


        + ======================================================================================== +
        |                                [  Food Class Methods  ]                                  |
        + ======================================================================================== +

    1)  new_food()
           +----------------------+---------------+
           | Parameter of Method  | Default Value |
           +----------------------+---------------+
           |     x_pos_border     |  275          |
           |     x_neg_border     |  -275         |
           |     y_pos_border     |  275          |
           |     y_neg_border     |  -274         |
           +----------------------+---------------+
           It Creates Food with random location and,
           It Increases or Decreases area food show

    2)  is_food_eat()
           +----------------------+---------------+
           | Parameter of Method  | Default Value |
           +----------------------+---------------+
           |     snake_object     |  User_def     |
           +----------------------+---------------+
           Return Boolean value
           It checks Food is Eaten and,
           It Makes new food


  ========================================================================================================


        + ======================================================================================== +
        |                              [  Scoreboard Class Methods  ]                              |
        + ======================================================================================== +

    1)  scoreboard_body()
           +----------------------+--------------------------+
           | Parameter of Method  |      Default Value       |
           +----------------------+--------------------------+
           |     alignment        |  "center"                |
           |     font             |  ('Arial', 20, 'normal') |
           |     coordinate       |  None                    |
           +----------------------+--------------------------+
           It Creates scoreboard Body and,
           It set position Separately of ScoreBoard [IF needed]
           Note :- If you change position use [ scoreboard_coordinates ] Variable from class

             eg :- Import SnakeGame
                   scoreboard = SnakeGame.Scoreboard()
                   scoreboard.scoreboard_coordinates = (180, 270)
                        [Any Coordinates In Tuple(x, y)]

    2)  increase_score()
            It increases points by 1

    3)  reset()
            It Reset all values to default or Set score to [zero]
            Note:- It Not reset HighScore

    4)  game_over()
           +----------------------+--------------------------+
           | Parameter of Method  |      Default Value       |
           +----------------------+--------------------------+
           |     text             |  GAME OVER!              |
           |     alignment        |  "center"                |
           |     font             |  ('Arial', 20, 'normal') |
           +----------------------+--------------------------+
            It shows GAME OVER! Message at Middle of Screen
            IT changes Text                  |
            It changes Alignment of Message  |-- [If needed]
            IT changes Font of Message       |

CHANGELOG.txt
Change Log
==========

0.1.1 (3/12/2021)
-------------------
- First Release

snake.py
from turtle import Turtle

start_point = [(0, 0), (-20, 0), (-40, 0), (-60, 0)]
segment_between_space = 15
UP = 90
DOWN = 270
RIGHT = 0
LEFT = 180
PIXEL = 10
SPEED = 0


class Snake:

    def __init__(self):
        self.snake_segment = []
        self.snake_body()
        self.head = self.snake_segment[0]
        self.head.shape("triangle")
        self.tail = self.snake_segment[2:]

    def snake_body(self):
        """Return Snake Segment object"""
        for pos in start_point:
            self.add_segment(pos)

    def add_segment(self, position,
                    shape="square",
                    color="black",
                    length=0.5,
                    width=0.5):
        """Add segment"""
        snakes = Turtle(shape=shape)
        snakes.color(color)
        snakes.turtlesize(stretch_len=length, stretch_wid=width)
        snakes.penup()
        snakes.goto(position)
        self.snake_segment.append(snakes)

    def extend(self):
        """ADD Segment at Last Segment Position"""
        self.add_segment(self.snake_segment[-1].position())

    def move(self):
        """Return Snake Head Position and
           Move Snake Forward"""
        for seg_num in range(len(self.snake_segment) - 1, 0, -1):
            x_coordinate = self.snake_segment[seg_num - 1].xcor()
            y_coordinate = self.snake_segment[seg_num - 1].ycor()
            self.snake_segment[seg_num].goto(x_coordinate, y_coordinate)
        self.head.forward(segment_between_space)
        return self.head.position()

    def right(self):
        """Move Snake Right"""
        if self.head.heading() != LEFT:
            self.head.setheading(RIGHT)

    def up(self):
        """Move Snake Up"""
        if self.head.heading() != DOWN:
            self.head.setheading(UP)

    def left(self):
        """Move Snake Right"""
        if self.head.heading() != RIGHT:
            self.head.setheading(LEFT)

    def down(self):
        """Move Snake Down"""
        if self.head.heading() != UP:
            self.head.setheading(DOWN)

    def reset(self):
        """Reset Snake"""
        for seg in self.snake_segment:
            seg.goto(1000, 1000)
        self.__init__()

    @staticmethod
    def snake_boundary(head_position,
                       x_positive=300,
                       x_negative=-300,
                       y_positive=300,
                       y_negative=-300):
        """Return Bool value"""
        if head_position[0] > x_positive or head_position[0] < x_negative or head_position[1] > y_positive \
                or head_position[1] < y_negative:
            return False
        else:
            return True


snakes.py
from turtle import Turtle

start_point = [(0, 0), (-20, 0), (-40, 0), (-60, 0)]
segment_between_space = 15
UP = 90
DOWN = 270
RIGHT = 0
LEFT = 180
PIXEL = 10
SPEED = 0


class Snake:

    def __init__(self):
        self.snake_segment = []
        self.snake_body()
        self.head = self.snake_segment[0]
        self.head.shape("triangle")
        self.tail = self.snake_segment[2:]

    def snake_body(self):
        """Return Snake Segment object"""
        for pos in start_point:
            self.add_segment(pos)

    def add_segment(self, position,
                    shape="square",
                    color="black",
                    length=0.5,
                    width=0.5):
        """Add segment"""
        snakes = Turtle(shape=shape)
        snakes.color(color)
        snakes.turtlesize(stretch_len=length, stretch_wid=width)
        snakes.penup()
        snakes.goto(position)
        self.snake_segment.append(snakes)

    def extend(self):
        """ADD Segment at Last Segment Position"""
        self.add_segment(self.snake_segment[-1].position())

    def move(self):
        """Return Snake Head Position and
           Move Snake Forward"""
        for seg_num in range(len(self.snake_segment) - 1, 0, -1):
            x_coordinate = self.snake_segment[seg_num - 1].xcor()
            y_coordinate = self.snake_segment[seg_num - 1].ycor()
            self.snake_segment[seg_num].goto(x_coordinate, y_coordinate)
        self.head.forward(segment_between_space)
        return self.head.position()

    def right(self):
        """Move Snake Right"""
        if self.head.heading() != LEFT:
            self.head.setheading(RIGHT)

    def up(self):
        """Move Snake Up"""
        if self.head.heading() != DOWN:
            self.head.setheading(UP)

    def left(self):
        """Move Snake Right"""
        if self.head.heading() != RIGHT:
            self.head.setheading(LEFT)

    def down(self):
        """Move Snake Down"""
        if self.head.heading() != UP:
            self.head.setheading(DOWN)

    def reset(self):
        """Reset Snake"""
        for seg in self.snake_segment:
            seg.goto(1000, 1000)
        self.__init__()

    @staticmethod
    def snake_boundary(head_position,
                       x_positive=300,
                       x_negative=-300,
                       y_positive=300,
                       y_negative=-300):
        """Return Bool value"""
        if head_position[0] > x_positive or head_position[0] < x_negative or head_position[1] > y_positive \
                or head_position[1] < y_negative:
            return False
        else:
            return True

src/food.py
from turtle import Turtle
from random import randint

PIXEL = 10


class Food(Turtle):

    def __init__(self):
        super().__init__()
        self.shape("circle")
        self.turtlesize(stretch_wid=0.5, stretch_len=0.5)
        self.penup()
        self.color("blue")
        self.speed("fastest")
        self.food_position = ()
        self.new_food()

    def new_food(self, x_pos_border=275,
                 x_neg_border=-275,
                 y_pos_border=275,
                 y_neg_border=-275):
        """Make Food and Set Food Position"""
        x = x_pos_border
        minus_x = x_neg_border
        y = y_pos_border
        minus_y = y_neg_border
        coordinates = (
            float(randint(minus_x, x)),
            float(randint(minus_y, y))
            )
        self.goto(coordinates)
        self.food_position = coordinates

    def is_food_eat(self, snake_object):
        """Return Bool value"""
        if snake_object.distance(self.food_position) < PIXEL:
            self.new_food()
            return True
        else:
            return False


scoreboard.py
from turtle import Turtle

ALIGNMENT = "center"
FONT = ('Arial', 20, 'normal')


class Scoreboard(Turtle):

    def __init__(self):
        super().__init__()
        self.hideturtle()
        self.color("white")
        self.penup()
        self.score = 0
        with open("../data.txt") as data:
            self.high_score = int(data.read())
        self.scoreboard_coordinates = (0.00, 270)
        self.scoreboard_body(coordinate=self.scoreboard_coordinates)

    def scoreboard_body(self, alignment=ALIGNMENT,
                        font=FONT,
                        coordinate=None):
        """Scoreboard Body"""
        self.clear()
        self.setposition(coordinate)
        self.write(arg=f"[ Score : {self.score}  HighScore : {self.high_score} ]", align=alignment, font=font)

    def increase_score(self):
        """Increase ScoreBoard"""
        self.score += 1
        self.scoreboard_body(coordinate=self.scoreboard_coordinates)

    def reset(self):
        """Reset ScoreBoard"""
        if self.score > self.high_score:
            self.high_score = self.score
            with open("../data.txt", mode="w") as data:
                data.write(f"{self.high_score}")
        self.score = 0
        self.scoreboard_body(coordinate=self.scoreboard_coordinates)

    def game_over(self, text="GameOver!",
                  alignment="center",
                  font=('Arial', 20, 'normal')):
        """GameOver Message"""
        self.home()
        self.write(arg=text, align=alignment, font=font)




__init__.py
from turtle import Turtle
from random import randint

start_point = [(0, 0), (-20, 0), (-40, 0), (-60, 0)]
segment_between_space = 15
UP = 90
DOWN = 270
RIGHT = 0
LEFT = 180
SPEED = 0
PIXEL = 10
ALIGNMENT = "center"
FONT = ('Arial', 20, 'normal')


class Snake:

    def __init__(self):
        self.snake_segment = []
        self.snake_body()
        self.head = self.snake_segment[0]
        self.head.shape("triangle")
        self.tail = self.snake_segment[2:]

    def snake_body(self):
        """Return Snake Segment object"""
        for pos in start_point:
            self.add_segment(pos)

    def add_segment(self, position,
                    shape="square",
                    color="black",
                    length=0.5,
                    width=0.5):
        """Add segment"""
        snakes = Turtle(shape=shape)
        snakes.color(color)
        snakes.turtlesize(stretch_len=length, stretch_wid=width)
        snakes.penup()
        snakes.goto(position)
        self.snake_segment.append(snakes)

    def extend(self):
        """ADD Segment at Last Segment Position"""
        self.add_segment(self.snake_segment[-1].position())

    def move(self):
        """Return Snake Head Position and
           Move Snake Forward"""
        for seg_num in range(len(self.snake_segment) - 1, 0, -1):
            x_coordinate = self.snake_segment[seg_num - 1].xcor()
            y_coordinate = self.snake_segment[seg_num - 1].ycor()
            self.snake_segment[seg_num].goto(x_coordinate, y_coordinate)
        self.head.forward(segment_between_space)
        return self.head.position()

    def right(self):
        """Move Snake Right"""
        if self.head.heading() != LEFT:
            self.head.setheading(RIGHT)

    def up(self):
        """Move Snake Up"""
        if self.head.heading() != DOWN:
            self.head.setheading(UP)

    def left(self):
        """Move Snake Right"""
        if self.head.heading() != RIGHT:
            self.head.setheading(LEFT)

    def down(self):
        """Move Snake Down"""
        if self.head.heading() != UP:
            self.head.setheading(DOWN)

    def reset(self):
        """Reset Snake"""
        for seg in self.snake_segment:
            seg.goto(1000, 1000)
        self.__init__()

    @staticmethod
    def snake_boundary(head_position,
                       x_positive=300,
                       x_negative=-300,
                       y_positive=300,
                       y_negative=-300):
        """Return Bool value"""
        if head_position[0] > x_positive or head_position[0] < x_negative or head_position[1] > y_positive \
                or head_position[1] < y_negative:
            return False
        else:
            return True


class Food(Turtle):

    def __init__(self):
        super().__init__()
        self.shape("circle")
        self.turtlesize(stretch_wid=0.5, stretch_len=0.5)
        self.penup()
        self.color("blue")
        self.speed("fastest")
        self.food_position = ()
        self.new_food()

    def new_food(self, x_pos_border=275,
                 x_neg_border=-275,
                 y_pos_border=275,
                 y_neg_border=-275):
        """Make Food and Set Food Position"""
        x = x_pos_border
        minus_x = x_neg_border
        y = y_pos_border
        minus_y = y_neg_border
        coordinates = (
            float(randint(minus_x, x)),
            float(randint(minus_y, y))
            )
        self.goto(coordinates)
        self.food_position = coordinates

    def is_food_eat(self, snake_object):
        """Return Bool value"""
        if snake_object.distance(self.food_position) < PIXEL:
            self.new_food()
            return True
        else:
            return False


class Scoreboard(Turtle):

    def __init__(self):
        super().__init__()
        self.hideturtle()
        self.color("white")
        self.penup()
        self.score = 0
        with open("data.txt") as data:
            self.high_score = int(data.read())
        self.scoreboard_coordinates = (0.00, 270)
        self.scoreboard_body(coordinate=self.scoreboard_coordinates)

    def scoreboard_body(self, alignment=ALIGNMENT,
                        font=FONT,
                        coordinate=None):
        """Scoreboard Body"""
        self.clear()
        self.setposition(coordinate)
        self.write(arg=f"[ Score : {self.score}  HighScore : {self.high_score} ]", align=alignment, font=font)

    def increase_score(self):
        """Increase ScoreBoard"""
        self.score += 1
        self.scoreboard_body(coordinate=self.scoreboard_coordinates)

    def reset(self):
        """Reset ScoreBoard"""
        if self.score > self.high_score:
            self.high_score = self.score
            with open("data.txt", mode="w") as data:
                data.write(f"{self.high_score}")
        self.score = 0
        self.scoreboard_body(coordinate=self.scoreboard_coordinates)

    def game_over(self, text="GameOver!", alignment="center", font=('Arial', 20, 'normal')):
        """GameOver Message"""
        self.home()
        self.write(arg=text, align=alignment, font=font)
